packages/query-store/src/QueryStore.ts

```diff
@@ -0,0 +1,340 @@
+		return this.#query.toString();
+	}
+
+	/** Name and Type information about the result columns */
+	#columns: ColumnMetadata[] = [];
+	get columns() {
+		return Array.from(this.#columns);
+	}
+
+	/** Has #fetchData been executed? */
+	#loaded = false;
+	get loaded() {
+		return this.#loaded;
+	}
+	/** Is #fetchData currently running? */
+	#loading = false;
+	get loading() {
+		return this.#loading;
+	}
+
+	/** Has #fetchLength been executed? */
+	#lengthLoaded = false;
+	get lengthLoaded() {
+		return this.#lengthLoaded;
+	}
+	/** Is #fetchLength currently running? */
+	#lengthLoading = false;
+	get lengthLoading() {
+		return this.#lengthLoading;
+	}
+
+	#length?: number;
+	get length(): number {
+		return this.#length ?? 0;
+	}
+
+	/**
+	 * Svelte bases iteration on the `length` property; so that has to exist before it will try to pass a number
+	 * However, we don't want to load everything if `length` is accessed - only the data itself.
+	 * Therefore, there is a tick when the array is full of undefined, because we are still fetching.
+	 * In that time period, we return this empty object so we don't get `cannot access x of undefined` errors.
+	 * This may break in instances where the user may have nested arrays / dicts
+	 */
+	#mockResult: QueryResult = {};
+
+	readonly id: string;
+
+	constructor(
+		query: string | Query,
+		exec: Runner,
+		id?: string,
+		private readonly opts: QueryStoreOpts = { disableCache: false },
+		private readonly root?: QueryStore
+	) {
+		super();
+		Object.freeze(opts);
+		// Ensure an ID Exists
+		if (!id) this.id = buildId(query);
+		else this.id = id;
+
+		// TODO: Strip any trailing ; from queries
+		// This is hard because of comments
+		// We might want to just error out if the querystring contains a ; for simplicity
+
+		if (typeof query === 'string') {
+			this.#query.from({ __userQuery: sql`(${query})` }).select('*');
+		} else this.#query = query;
+		// @ts-expect-error Passing undocumented parameter
+		this.#exec = (...args: Parameters<Runner>) => exec(args[0], args[1] ?? this.id);
+
+		this.#proxied = new Proxy<QueryStore & QueryResult[]>(
+			this as unknown as QueryStore & QueryResult[],
+			{
+				get: (self, _prop) => {
+					// Intercept numeric indices. This implies we're trying to access rows (data) in the store.
+					// If the data has not been loaded, initiate the async #update method to fetch the data.
+					let prop: string | symbol | number = _prop;
+					if (typeof prop === 'string' && /^[\d.]+$/.exec(prop)) prop = parseInt(prop);
+					if (typeof prop === 'number') {
+						if (!self.#loaded) {
+							try {
+								const r = self.#fetchData();
+								if (r instanceof Promise)
+									r.catch((e) => {
+										throw new Error('Failed to update query store', { cause: e });
+									});
+							} catch (e) {
+								throw new Error('Failed to update query store', { cause: e });
+							}
+						}
+
+						if (!self.#values[prop]) return self.#mockResult;
```

Comment:
Was that added for `testing`?

Reply:
`#mockResult` is an empty object that contains keys based on the rows returned from the metadata query; this prevents errors coming from `cannot read property "x" of undefined` when accessing `query[0].x` before the data has been fetched

Reply:
See also:

```typescript
	/**
	 * Svelte bases iteration on the `length` property; so that has to exist before it will try to pass a number
	 * However, we don't want to load everything if `length` is accessed - only the data itself.
	 * Therefore, there is a tick when the array is full of undefined, because we are still fetching.
	 * In that time period, we return this empty object so we don't get `cannot access x of undefined` errors.
	 * This may break in instances where the user may have nested arrays / dicts
	 */
	#mockResult: QueryResult = {};
```